### PR TITLE
[SER-515] Division of Data Manager CRUD rights

### PR DESCRIPTION
**Pull request by John-Weak**
[feature] multiple permsissions with inheritance/dependency

### CODE REVIEW

File: api/utils/requestProcessor.js [hunk 1]
```diff
@@ -1424,8 +1424,26 @@ const processRequest = (params) => {
                 case 'permissions':
                     validateRead(params, 'core', function() {
                         var features = ["core", "events" /* , "global_configurations", "global_applications", "global_users", "global_jobs", "global_upload" */];
-                        plugins.dispatch("/permissions/features", {params: params, features: features}, function() {
-                            common.returnOutput(params, features);
+                        /*
+                            Example structure for featuresPermissionDependency Object
```
Comment (kanwarujjaval) [inline api/utils/requestProcessor.js:1428]: Can you add other keys too
not just 'c'.
example should be more comprehensive that generally covers as many cases as possible for an implementation

File: bin/upgrade/DEV/upgrade_user_permission_data-manager.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
+var pluginManager = require('../../../../plugins/pluginManager.js')
```
Comment (kanwarujjaval) [inline bin/upgrade/DEV/upgrade_user_permission_data-manager.js:1]: Please add a brief documentation as to what this script does.

File: frontend/express/public/core/user-management/javascripts/countly.models.js [hunk 3]
```diff
@@ -142,7 +152,40 @@
                 _t: Date.now()
             },
             success: function(res) {
-                _features = res;
+                _features = res.features;
```
Comment (kanwarujjaval) [inline frontend/express/public/core/user-management/javascripts/countly.models.js:155]: Add brief documentation (comments) for this

File: frontend/express/public/core/user-management/javascripts/countly.views.js [hunk 4]
```diff
@@ -386,9 +394,66 @@
                 this.permissionSets.splice(index, 1);
                 this.$set(this.$refs.userDrawer.editedObject.permission._.u, this.$refs.userDrawer.editedObject.permission._.u.splice(index, 1));
             },
+            setPermissionByDependency: function(index, type, feature) {
```
Comment (kanwarujjaval) [inline frontend/express/public/core/user-management/javascripts/countly.views.js:397]: Does this belong in the views.js or can we turn this into an action in the models file?

File: frontend/express/public/javascripts/countly/countly.helpers.js [hunk 5]
```diff
@@ -12,13 +12,28 @@
 (function(CountlyHelpers) {
 
     CountlyHelpers.isPluginEnabled = function(name) {
-        if (countlyGlobal && countlyGlobal.pluginsFull && Array.isArray(countlyGlobal.pluginsFull) && countlyGlobal.pluginsFull.indexOf(name) > -1) {
-            if (countlyGlobal.plugins && Array.isArray(countlyGlobal.plugins) && countlyGlobal.plugins.indexOf(name) !== -1) {
-                return true;
+        if (countlyGlobal && countlyGlobal.pluginsFull && Array.isArray(countlyGlobal.pluginsFull)) {
```
Comment (kanwarujjaval) [inline frontend/express/public/javascripts/countly/countly.helpers.js:15]: Please document these changes**
This function is missing jsdoc, can we add it ?

File: frontend/express/public/javascripts/countly/vue/core.js [hunk 6]
```diff
@@ -83,20 +83,31 @@
     };
 
     var authMixin = function(featureName) {
+        if (!Array.isArray(featureName)) {
```
Comment (kanwarujjaval) [inline frontend/express/public/javascripts/countly/vue/core.js:86]: Please document these changes

**Review by kanwarujjaval — CHANGES_REQUESTED** [review]

File: frontend/express/public/core/user-management/javascripts/countly.views.js (hunk 4 above)
Comment (John-Weak) [reply frontend/express/public/core/user-management/javascripts/countly.views.js:403]: all the similar logic ex: automatic read permission, also exists in view.js hence I decided to keep similar things together 😀

Comment (kanwarujjaval) [reply frontend/express/public/core/user-management/javascripts/countly.views.js:403]: Okay 👍
We should refactor this once the changes are tested and stable
FYI @widatama

**Review by kanwarujjaval — APPROVED** [review]

**Pull request merged**